java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,122 @@
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** A representation of a lookup claim in a Paxel [Expression]. */
+  data class Equal(val path: Path = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Return [Equal] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Equal = Equal(
+      path.flatMap { identifier ->
+        when (val association = aliases.associations.getOrDefault(identifier, Equal(identifier))) {
+          is Equal -> association.path
+          is Derive -> association.firstPath().path
+          is Scope -> throw UnsupportedOperationException(
+            "Cannot substitute a Scope for an Identifier."
+          )
+        }
+      }
+    )
+
+    /** Union of a [Equal] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(this.path) + Derive(other.path)
+      is Derive -> Derive(this.path) + other
+      is Scope -> throw UnsupportedOperationException(
+        "Union of Equal and Scope is not well defined."
+      )
+    }
+  }
+
+  /** A representation of a Derivation claim in a Paxel [Expression]. */
+  data class Derive(val paths: List<Path> = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Path) : this(paths.toList())
+
+    /** Union of a [Derive] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> this + Derive(other.path)
+      is Derive -> Derive(paths + other.paths)
+      is Scope -> throw UnsupportedOperationException(
+        "Union of Derive and Scope is not well defined."
+      )
+    }
+
+    /** Returns the first [Equal] in the collection, or an empty [Equal]. */
+    fun firstPath(): Equal = if (paths.isNotEmpty()) Equal(paths.first()) else Equal()
+
+    /** Substitute all aliases as a new [Derive] object. */
+    override fun substitute(aliases: Scope) = Derive(
+      paths.map { path -> Equal(path).substitute(aliases).path }
+    )
+  }
+
+  /** Associates an [Identifier] with a group of [Deduction]s. */
+  data class Scope(
+    val associations: Map<Identifier, Deduction> = emptyMap()
+  ) : Deduction() {
+
+    constructor(vararg pairs: Pair<Identifier, Deduction>) : this(pairs.toMap())
+
+    /** Returns a new [Scope] with only the queried association. */
+    fun lookup(key: Identifier): Scope {
+      require(key in associations) {
+        "Scope does not associate anything with '$key'."
+      }
+      return Scope(key to associations.getOrDefault(key, Derive()))
+    }
+
+    /** Union of a [Scope] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> this + Derive(other.path)
+      is Derive -> this + other
+      is Scope -> Scope(
+        associations = (this.associations.entries + other.associations.entries)
+          .fold(emptyMap()) { acc, (key, value) ->
```

Comment:
Is there a more specific kotlin function that I could use here?

Reply:
Maybe:
```
(this.associations.asSequence() + other.associations.asSequence())
 .groupBy({ it.key }, { it.value })
 .mapValues((_, values) -> when (values.size) {
    1: values[0]
    2: values[1] + values[2]
    else -> throw IllegalStateException(...)
  } )
```

Reply:
I kind of prefer a `fold` over using two HOFs. I could be convinced otherwise :)

Reply:
If you do use fold, please modify it to use a mutable map instead of constructing a new map instance with each iteration.